### ISSUE TITLE
Discussion: Move built-in packages to a separate repository

### ISSUE THREAD

**Issue by tldahlgren** [LLNL/spack#47480]
Supersedes https://github.com/spack/spack/issues/1773

@tgamblin @becker33 @haampie @alalazo @psakievich @citibeth

<a name="contents"></a>
### Contents

- [Summary](#summary)
- [Task Checklist](#checklist)
- [Rationale](#rationale)
- [Description](#description)
  - [Affected Paths](#affected-paths)
  - [PR Checks](#pr-checks)
  - [Bootstrapping](#bootstrapping)
  - [Package Repository](#package-repository)
  - [Repository Updates and Synchronizations](#repository-maintenance)
  - [Package Commands](#package-commands)
- [Approach](#approach)
- [Additional Information](#Additional-Information)
- [General Information](#General-Information)
 
<a name="summary"></a>
## Summary

The Spack repository contains a significant number of built-in package recipes and associated files that are needed for installing software. This discussion proposes to move those files into a separate repository and modify the core software to support the change.

<a name="checklist"></a>
## Task Checklist

The following is a checklist of the high-level tasks described in [Approach](#approach). Each task should correspond to an approach subsection.

- [x] [Explore Relevant Features of Other Products](#related-work) (@alalazo, @scheibelp)
- [x] [Document User Stories or Use Cases](#user-stories) (@scheibelp, @becker33, @zackgalbreath )
- [x] [Move Build Systems into BuiltIn](#build-systems) (@haampie)
- [x] [Assess Untangling Unit Tests](#unit-tests) (@tldahlgren)
- [x] [Untangle Compiler Wrappers](#compiler-wrappers) (@alalazo)
- [x] [Support Repository Compatibility](#repo-compatibility) (@becker33, @zackgalbreath )
- [x] [Determine Support for Bootstrapping](#support-bootstrapping) (@becker33)
- [x] [Assess Untangling GitHub Actions and CI](#pr-actions-ci) (@zackgalbreath/Kitware)
- [x] [Separate the Builtin Repository](#separate-builtin-repo) (@zackgalbreath/Kitware, @haampie)
- [x] [Support and Access Package Repositories](#package-repos) (@haampie )
- [ ] [Confirm Package-Related Commands Work](#package-commands-work)
- [ ] [Add Repository CLI Commands](#repo-cli) .. tied to use cases; defer to step 2

Back to [Contents](#contents)

<a name="rationale"></a>
## Rationale

Spack packages specify the versions, options, and processes required for installing and optionally testing software. Their directories may contain additional files, such as patches and custom tests. For some, the number of associated files is a significant hurdle while package drift is a key issue for others. This change is also considered by the community as a requirement of Spack 1.0.

With over 8,340 packages, the number of files associated with built-in packages currently comprise 86% of a fresh clone of the Spack repository. Embedding the ever growing number of packages in the same repository as the core software creates a burden for some installation environments, especially those with space and or i-node constraints.

Another problem with maintaining built-in packages in the same repository is that packages generally evolve independently, which can be problematic for projects that rely on a specific version (or commit) of a package (e.g., for software they've already installed) but want or need to get updates to the core Spack software. This "drift" in the package implementation can lead to the unanticipated and possibly undesirable installation of new versions of dependent software.

Maintaining packages in the same repository as the core software leads to problems for many Spack users. These problems range from excessive use of file space/i-nodes to the effects on project development associated with package implementation "drift".

Back to [Contents](#contents)

<a name="description"></a>
## Description

While Spack already supports multiple package repositories (https://spack.readthedocs.io/en/latest/repositories.html), there are additional considerations beyond simply moving the built-in packages folder into a separate repository (https://docs.github.com/en/get-started/using-git/splitting-a-subfolder-out-into-a-new-repository) and configuring the new repository. Hence, the need for discussing issues here include:

- Which directories and files need to be moved?
- How will PR checks be affected?
- How will spack fetch packages for bootstrapping? Where?
- What is the process for fetching and configuring the package repository?
- How will updating separate repositories be managed and changes synchronized?
- What changes, if any, are needed to the many commands that work on the built-in repository?

Back to [Contents](#contents)

<a name="affected-paths"></a>
### Affected Directories and Files

Which directories and files need to be moved?

At a minimum the new package repository shall contain the build systems (`$SPACK_ROOT/lib/spack/spack/build_systems`) and built-in packages (under `$SPACK_ROOT/var/spack/repos/builtin/packages`). Additional files may need to be moved depending on the approach chosen for handling CI checks (described in the next section).

Mock packages are **_not_** to be moved to the new repository since they are integral to Spack's unit tests.

Back to [Contents](#contents)

<a name="pr-checks"></a>
### PR Checks

How will PR checks be affected?

Pull requests (PRs) trigger a number of checks related to packages. One of the checks uses GitLab CI pipelines to rebuild modified packages to maintain the build cache. Our plan is to  migrate these pipelines to the new packages repository (away from core Spack). Unit tests in the core Spack repo will be responsible for properly exercising the package API to maintain compatibility between these two repos.

Spack performs additional checks to include style and audits. Those relevant to packages-only would need to be moved to the new repository; while common ones would need to be copied.

At some point in the not-too-distant future we hope to trigger running stand-alone tests for modified packages. This check could continue as originally envisioned to be part of the CI workflow approach above or could be separated (assuming it can be triggered *after* any build cache updates or any non-build cache builds that might be added by @bernhardkaindl as discussed in the 2024 Nov 6 Technical Steering Committee meeting).

Back to [Contents](#contents)

<a name="bootstrapping"></a>
### Bootstrapping

How will spack fetch packages for bootstrapping? Where?

Spack has several sets of requisite software that must be bootstrapped:

- `clingo` (https://spack.readthedocs.io/en/latest/getting_started.html#bootstrapping-clingo);
- `gnupg`; and
- `patchelf`.

Back to [Contents](#contents)

<a name="package-repository"></a>
### Package Repository

What processes will be used for fetching and configuring the package repository?

Spack will need the ability to configure the package repository and a process for fetching packages. Spack currently supports multiple package repositories through entries in a `repos.yaml` file (see the docs at https://spack.readthedocs.io/en/latest/repositories.html). However, those entries are required to be on the file system.

If that mechanism is retained, then it will need to support remote package repositories (specified with URLs that can include a commit hash) and the default configuration (`$SPACK_ROOT/etc/spack/defaults/repos.yaml`) modified to point to the new package repository. The relevant packages would need to at least be fetched from that location and likely cached locally.

Alternatively, spack could support a package registry with a modified or alternate representation for the configuration (such as that used by `cargo`, which supports named registries with URLs).

Progress:
- New repository: https://github.com/spack/spack-packages
- https://github.com/spack/spack/pull/50322

Back to [Contents](#contents)

<a name="repository-maintenance"></a>
### Repository Updates and Synchronization

How will updating separate repositories be managed and changes synchronized?

There are changes to Spack core that require changes to packages and or, if moved, CI stacks. Some features, such as adding support for stand-alone tests, are optional while others, such as a change in syntax (e.g., maintainers) or the schema for configuring CI stacks, are not. Mandatory changes to packages resulting from changes to the core require coordination of updates to both repositories.

One approach to managing updates and synchronization is to include the package repository as a git submodule of the Spack (core) repository. While this approach makes it somewhat easier to synchronize changes to both repositories (since you can clone the project repository under the core repository), its use of git is more advanced than many contributors may be accustomed (https://git-scm.com/book/en/v2/Git-Tools-Submodules).

An alternative is to rely on some form of dependency management. (TBD: What would this look like? Would we support synchronization features related to the repository configuration? CI checks?)

Back to [Contents](#contents)

<a name="package-commands"></a>
### Package Commands

What changes, if any, are needed to the many commands that work on the built-in repository?

Commands that expect to operate on built-in packages shall continue to do so, which we'll need to be confirmed (https://spack.readthedocs.io/en/latest/command_index.html#command-reference). These include package creation (e.g., `create`), query (e.g., `find`, `info`), build (e.g., `gc`), and developer (e.g., `blame`, `pkg`).

<a name="approach"></a>
## Approach

There is a significant amount of work needed to accomplish a well designed, implemented, and supported split for the repository. We split the work into more manageable tasks. The high-level tasks, not in priority order, are described here.

<a name="related-work"></a>
### Explore Relevant Features of Other Products

There are several tools having relevant features that we would like explored in terms of processing. These can include:
- homebrew (https://github.com/Homebrew/brew)
- Nix (https://github.com/nixos)
- pixi (https://github.com/prefix-dev/pixi)
- rye (https://github.com/astral-sh/rye)
- uv (https://github.com/astral-sh/uv)

And possibly:
- apt (https://salsa.debian.org/apt-team/apt)
- cargo (https://github.com/rust-lang/cargo)
- npm (https://github.com/npm)

Back to [Contents](#contents)

<a name="user-stories"></a>
### Document User Stories or Use Cases

It would be helpful to everyone working on these tasks to have a better understanding of the use cases this work is addressing, to include the uses for and how people expect to utilize the features being implemented here.

#### Anticipated Use Cases
**1. pinned Spack + pinned Packages**
  - _Intended audience_: users that greatly value stability and are okay using older versions of software.
  - _Pros_:
    - Most stable
  - _Cons_:
    - Cannot install new versions of packages as they are released
    - Cannot benefit from new features and performance improvements as they are added to Spack
  - _CI requirements_:
    - one-time thorough installation testing (and build cache population) for packages of interest when releases are created.

**2. pinned Spack + floating Packages**
  - _Intended audience_: users that value stability but want to keep their dependencies as up-to-date as possible.
  - _Pros_:
    - Can quickly and easily install new versions of packages.
    - Minimize exposure to temporary regressions in the core Spack code base.
  - _Cons_:
    - “Early adopter” for packages: willing to accept temporary regressions in package definitions.
  - _CI requirements_:
    - CI pipelines using a recently verified version of Spack to test relevant, proposed changes to the Packages repository before they are merged. This is essentially our existing GitLab CI pipelines, which will be relocated to the Packages repo.

**3. floating Spack + pinned Packages**
  - _Intended audience_: users with pinned dependency versions that are eager to adopt new functionality and performance improvements for Spack.
  - _Pros_:
    - Access to latest Spack features & improvements
    - Stable package recipes
  - _Cons_:
    - Can’t easily install new versions of dependencies
    - Willingness to accept temporary regressions in core Spack
  - _CI requirements_:
    - Unit tests in Spack that thoroughly exercise the Package API.
    - More generally, Spack's existing pytest suite.

**4. floating Spack + floating Packages**
  - _Intended audience_: users that want cutting edge dependency versions and the latest features Spack has to offer.
  - _Pros_:
    - access to both the latest features in Spack, as well as most up-to-date versions of package recipes.
  - _Cons_:
    - willingness to accept some instability as regressions are addressed in either repo.
  - _CI requirements_:
    - This use case will automatically benefit from CI supporting use cases **2)** and **3)**.
    - Will additionally require acceptance tests where we periodically (weekly?) do a “rebuild everything” with bleeding edge versions of both repos.
    - If these acceptance tests pass, we bump the pinned version of Spack used for testing in the packages repo (and vice versa, if we end up deciding to test some packages from core Spack).
    - When acceptance testing fails we will try to fix it quickly to resume automated snapshot releases.

Back to [Contents](#Contents)

<a name="build-systems"></a>
### Move Build Systems into BuiltIn Package Repository

Determine what needs to be moved from the core's build systems and into the **builtin** package repository. This task will include any restructuring and refactoring of existing modules and updates/additions to the corresponding unit tests.

Since these changes will impact how packages utilize the core's API, packages that violate the new API will need to be identified and their upgrade status tracked. One proposed option is to maintain a file the records individual packages that violate the new API. One example that was given was `dbcsr`, which utilizes the protected (some say private) function  `self._if_ninja_target_execute()`.

See https://github.com/spack/spack/issues/47480#issuecomment-2514894170 for a snapshot of imports that could be affected by this task.

#### Progress
-  #48760   
-  #48765  
- #48793 
- #48792 

Back to [Contents](#contents)

<a name="unit-tests"></a>
### Assess Untangling Unit Tests

Determine what aspects of unit tests are tied to packages in the **builtin** package repository (e.g., testing ecosystem) and which are specific to the core software. Come up with solutions to any issues such that the (remaining/revised) core unit testing process works after the **builtin** package repository is moved to a new GitHub repository.

Make as many changes as possible prior to the actual repository spit and create a plan for the remaining tasks.

#### Progress
~- #48232  (Umbrella PR)~
- #48926   
- #48927  
~- #48930~ .. has since been split into multiple independent PRs
- Decide what to do with symlinks from `builtin.mock` to `builtin` see https://github.com/spack/spack/pull/45189#discussion_r1986667018 (see https://github.com/spack/spack/pull/50478)

Back to [Contents](#contents)

<a name="compiler-wrappers"></a>
### Untangle Compiler Wrappers

Compiler wrappers are part of the package hash so this work needs to be assessed to determine what should be part of the core and what should be move to the separate **builtin** package repository. Implement solutions to any issues that arise, which may include restructuring and refactoring existing modules (e.g., establishing a compiler interface in Spack core) and updates/additions to the corresponding unit tests.

- [x] Move the `compiler-wrapper` into its own package
- [ ] Try to make the compiler wrapper less coupled to Spack internals. In #45189 the `compiler-wrapper` package _does not_ depend on Spack's configuration, but it still sets a few environment variables that are _needed_ by Spack. See discussion in https://github.com/spack/spack/pull/45189#discussion_r1990568305

**Warning**: This task is dependent on maturing support for compilers as dependencies.

Back to [Contents](#contents)

<a name="repo-compatibility"></a>
### Support Repository Compatibility

After the split, we anticipate maintaining the following independent versions:

* Spack version (semver, `major.minor.patch`)
  - We expect to maintain the current cadence of two releases per year.
* Spack package API version (semver, `major.minor`)
  - The package API includes the `spack.package` module, which is a small subset of Spack's scripting API. It also includes the structure of a repository (both on the filesystem and as Python modules).
  - The minor version is bumped if the package API is extended in a backward compatible way. For example: a new directive is added.
  - The major version is bumped if the package API has a breaking change. For example: a directive is removed, a directive's function signature is changed in a breaking way, or the repository filesystem layout is changed in an incompatible way. We will strive to avoid major version bumps as much as possible.
  - Each package repo will define the package API version for recipes contained within. If a repo specifies compatibility with package API version `1.3` it means `>=1.3` and `<2`.
  - CI for the package repo will need to use an appropriate version of Spack (one that can understand this package API version).
* Spack package repo version (date + patch release, e.g. `2025-06.2`)
  - Used for binary caches

#### Compatibility Guarantees

All versions of Spack within a major release stream (e.g. `1.x.y`) will be able to understand the package API version supported by the initial major release (`1.0.0`). If at some point in the future Spack drops support for a package API version, this will require a major version bump for Spack.

Every Spack release will be able to read packages for the prior **two years** worth of Package Repo releases.

When a backward incompatible change to the package API is deemed necessary, we will release a final supporting version of Spack and the official Package repo before bumping the package API version.

Spack will define and announce what Package API version(s) it understands.

Back to [Contents](#contents)

<a name="support-bootstrapping"></a>
### Determine Support for Bootstrapping

As described in Bootstrapping](#bootstrapping), Spack relies on a core set of packages. Determine what is needed to support bootstrapping once the **builtin** package repository is no longer available.

Back to [Contents](#contents)

<a name="pr-actions-ci"></a>
### Assess Untangling GitHub Actions and CI

As discussed in [PR Checks](#pr-checks), Spack has GitHub actions that result in a variety of checks against the proposed changes. Determine which checks are common to the core and package repositories, specific to the core software, and specific to the **builtin* package repository. Document where each belongs and what is needed to ensure they are being performed for the proper repositories.

Back to [Contents](#contents)

<a name="separate-builtin-repo"></a>
### Separate the Builtin Repository

Once the groundwork is in place in terms of the API and a plan is in place regarding GitHub Actions and CI, create the new **builtin** repository and configure the GitHub Actions and CI accordingly.  Be mindful of the [Affected Paths](#affected-paths).

Ensure that the build (and test) outputs continue to get reported in CDash at https://cdash.spack.io/index.php?project=Spack+Testing.

Progress:
- https://github.com/spack/spack-packages
- https://github.com/spack/spack/pull/50322

Back to [Contents](#contents)

<a name="package-repos"></a>
### Support and Access Package Repositories

One of the goals is to support the **builtin** package repository in the same manner as other package repositories. So there will be no use of git submodules or subsets. This means that not only will **builtin** packages be moved into and retrieved from a separate (GitHub) repository, but the mechanisms for staging, ensuring compatibility, moving package repositories to their local (cache) location, and accessing the packages from the local cache need to be designed and implemented.

Spack already supports caching of included remote files for environments but the goal is to design and implement the processes such that they can be used to support not only package repositories but mirrors and environments.

Preliminary ideas for specifying repositories have the following forms:

`repos.yaml`:
```
repos:
- url: https://github.com/my/package/repo.git
  ref: 1.0.0
  namespace: builtin
  path: /path/to/local/repo
```

Repositories can also be overridden in a `spack.yaml` file:
```
spack:
  repos::
  - url: https://github.com/other/package/repo.git
    ref: develop
    namespace: bultin
    path: /path/to/local/builtin/repo
```

If a path is not given, then it would default to one specified/calculated by Spack.

See also Package Repository](#package-repository).

**Warning**: This task relies on the work done in [Support Repository Compatibility](#repo-compatibility).

Back to [Contents](#contents)

<a name="package-commands-work"></a>
## Confirm Package-Related Commands Work

Confirm that all of the package-related commands continue to work as before with the separated **builtin** package repository.

Back to [Contents](#Contents)

<a name="repo-cli"></a>
## Add Repository CLI Commands

Design and implement new CLI (sub)commands for interacting with the repositories. At a minimum, there should be the ability to retrieve and update repositories. Preliminary ideas include:

```
$ spack repo get  [<repo-name>]  # <repo-name> assumes presence in `repos.yaml`; get all repos if no name provided
$ spack update <repo-name> [<version>]   # Update the named repository to the specified version
$ spack update --advance-packages  # Alternate proposal to update spack *and* get the latest package repo(s) commits
```

In the case of the first syntax for `spack update`, the `<version>` is optional because a repo with a reference to a branch (e.g., `develop`) would only need to retrieve the latest commits.

See https://github.com/spack/spack/issues/47480#issuecomment-2513423648 for the alternate `spack update` syntax.

The `update` command would need to update `repos.yaml` if the version is different from that provided. The command should also ensure a new version is compatible with Spack's version.

Back to [Contents](#Contents)

<a name="additional-information"></a>
## Additional information

- Cargo registries: https://doc.rust-lang.org/cargo/reference/registries.html
- Git submodules: https://git-scm.com/book/en/v2/Git-Tools-Submodules
- Spack's package repository configuration: https://spack.readthedocs.io/en/latest/repositories.html

Back to [Contents](#Contents)

<a name="general-information"></a>
## General information

- [X] I have searched the issues of this repo and believe this is not a duplicate of an open issue

Back to [Contents](#contents)

**Comment by scheibelp** [comment]
Rough proposal

- Instead of updating Spack with `git pull`, we have `spack update`
    - The builtin repo is maintained separately, e.g. as a git submodule
    - `spack update --advance-packages` will move the package repo forward and pull the latest spack
    - Point being you can choose whether to do that or not: if you want the latest spack without the latest packages, then you can get that
- The repo stores its version as metadata
    - and package_base.py also stores the package version
    - If these don't match, then we've detected a problem (and `spack update` could e.g. give you the latest spack core that is compatible with the builtin-1.0 package repo)
- Overall:
    - This allows us to keep the repos separate
    - but also, it allows us to make backwards-incompatible changes if we want

**Comment by haampie** [comment]
Using

```
python3 -c 'import isort, sys, collections; print("\n".join(f"{k}: {v}" for k, v in collections.Counter(x.module for f in sys.argv[1:] for x in isort.find_imports_in_file(f)).most_common()))' $(find var/spack/repos/builtin/packages/ lib/spack/spack/build_systems/ -name '*.py')
```

<details>

<summary>There are roughly 700+ import statements for ~63 internal modules, including those from the build systems:</summary>

```

spack.directives: 63
spack.build_systems.cmake: 52
llnl.util.tty: 51
llnl.util.filesystem: 49
spack.build_systems: 42
spack.util.environment: 41
spack.build_systems.autotools: 36
spack.build_systems.python: 34
spack.package_base: 29
spack.build_environment: 22
spack.builder: 22
llnl.util.lang: 18
spack.build_systems.makefile: 17
spack.phase_callbacks: 16
spack.operating_systems.mac_os: 14
spack.build_systems.meson: 13
spack.variant: 13
spack.error: 12
spack.multimethod: 12
spack.util.prefix: 11
spack.util.executable: 11
spack.spec: 10
spack.build_systems.generic: 10
spack.compiler: 9
spack.version: 9
llnl.util: 8
spack.platforms: 8
spack.compilers: 7
spack.hooks.sbang: 7
spack.package_test: 7
llnl.util.symlink: 5
spack.store: 5
archspec: 5
spack.build_systems.msbuild: 4
spack.platforms.cray: 4
spack.build_systems.nmake: 4
spack.util.module_cmd: 4
spack.install_test: 4
spack.util.url: 4
spack.tengine: 3
spack.util.elf: 3
spack.paths: 3
spack.config: 3
xml.etree.ElementTree: 2
spack.build_systems.go: 2
macholib: 2
llnl.util.link_tree: 2
spack.url: 2
spack.build_systems.lua: 2
spack.detection: 2
spack.deptypes: 2
spack.main: 1
spack.build_systems.cargo: 1
spack.user_environment: 1
spack.util.windows_registry: 1
spack.fetch_strategy: 1
archspec.cpu: 1
spack.util.libc: 1
spack.operating_systems.linux_distro: 1
spack: 1
spack.relocate: 1
spack.repo: 1
spack.util.path: 1
```
</details>

**Comment by scheibelp** [comment]
Use cases:

1. (rephrasing of example from rationale section) User builds a large DAG; they encounter a Spack bug; we fix it in core; they want the fix in core without changing all their packages.
    - For the most part, I assume users will want to advance Spack's core logic and freeze the package definitions. ~I can't think of a case where someone would want to freeze the core logic and advance the package definitions~ (edit: a case for this is provided at https://github.com/spack/spack/issues/47480#issuecomment-2516177660)
    - This is the main reason to split off the packages into a repo: when you do that, you get this for "free" and then introduce complexities (e.g. the following use cases)
        - I say "main" because IMO this is more complex to handle than controlling inodes: you can satisfy that use case without running into API-consistency problems described below
2. After separating out the packages, let's say we need to make a major change that affects many packages
    - e.g. we've spoken about requiring virtuals to have an explicit package, what if we didn't think of that until after separating out the package repo?
    - For this, the package repo and the spack core both need their own version for describing compatibility. Broadly speaking, for package-repo-version X and spack-core version Y, this can be broken down into the following possibilities:

      a. We offer something new in Spack core that packages can take advantage of: this means that `X+1` requires `>= Y+1)`
      b. We remove something in Spack core (possibly replacing it with something else): this means that `X` requires `<= Y`
3.  After separating out the packages, let's say we discover one of them is referencing an internal property (e.g. let's say we miss a reference in `dbcsr` to `self._if_ninja_target_execute`); we decide to refactor it: we add a new public property and we update `dbcsr` to use that
    - alternative scenarios:
      a. we might not have noticed the reference in `dbcsr` until after we removed `_if_ninja_target_execute`
      b. In example (a) we are referring to an attribute on the package, but a user might also access a property by "reaching into" objects that we provide to the user as part of the API, e.g. all phases get `(Spec, prefix)` as arguments, and could access `Spec.some_attribute_we_didnt_mean_to_expose` (it's hard to distinguish what those are, because we advertise being able to set arbitrary attributes on Specs as a means of inter-spec communication, e.g. `mpicc`).
    - [note 1] It may seem that moving build systems into the package repo would avoid this, but a similar problem would come up if a user maintains a separate repo that inherits from the base package repo build system: we would change `_if_ninja_target_execute` and suddenly their downstream package would be incompatible
    - Key point is: we don't want an incompatibility in one package to force us to say that the entire package repo is incompatible. The Spack/package-repo versioning is to help us delineate large (affecting many packages), intentional incompatibilities
    - We should have something more fine-grained that indicates that just that package is invalid for some small range of Spack commits

Proposed solutions

3. `spack update` could simply be smart enough to move both spack core and the packages repo past the range of commits where there is an incompatibility (edited)
    - ... or, we could raise an error only when a concretization result includes this package

**Comment by davidedelvento** [comment]
> For the most part, I assume users will want to advance Spack's core logic and freeze the package definitions.

I see some people wanting to do that, but... IMHO....

> I can't think of a case where someone would want to freeze the core logic and advance the package definitions

... I think that most people (and certainly myself) would want to do exactly that. Why? I am running spack version X (and I mean actually version X, put ourselves in the future and say `v1.3.1` not commit `abcdef3514afafzv`). In production on my cluster with many users. Can't take chances of "breaking" it. That version of spack provides LLVM `v27.6.4`. Since LLVM has a very fast development cycle, well before spack upgrades to `v1.4` LLVM releases versions 28, 29, 30 and 31. User comes and wants all of them. I don't want to upgrade whole spack and risk breaking the whole production system (and I don't have the immediate time availability to do the more labor intensive effort to do the upgrade and test everything), especially because in production I want to run only released software not "random" HEAD commits available at the moment.
So I'm stuck and can't provide what the user asks. To resolve this problem, maybe that's not the ideal workflow, but what I do currently, I manually backport that single package (and sometimes its dependencies as needed) in the current version of spack and install. That has the side benefit that down the road I can do a `git status` and a `git diff` to immediately show what is different so I don't have to manually track/document what I have changed. With this separation of repos, I would advance the package repo only (keeping the core spack repo frozen) and attempt the install with the previously-installed-on-the-cluster version of spack. Of course this works assuming the package repo has a faster release cycle compared to spack itself, otherwise I'd simple continue to manually backport. And of course this works if this independent advancement is supported by some mechanism which says "with spack version X you can move the package versions in range Y to Z.

**Comment by marcmengel** [comment]
Many of the issues discussed above can be resolved by slightly expanding the definition
of bootstrapping.    Since Spack currently needs to be bootstrapped anyhow, bootstrapping 
could also install 

-   recipe repositories and/or 
-   spack extension repositories; 

 from a minimum of a "base" repository of exactly those packages needed for bootstrapping, 
to one or more groups/clusters of recipe repositories, 
and extensions, including those maintained by other organizations. 
   
This behavior could be based on some sort of bootstrap config file,  which could be fetched via URL. 
You could then have an organization-standard bootstrap config which would include your organization's local
recipe repositories, and spack extensions, as well as a "standard"/default configuration which installs what
are now the builtins. 

CI configurations could include a bootstrap config which would ensure the mix of recipes desired for the tests in that
configuration.

**Comment by marcmengel** [comment]
@scheibelp wrote:
> After separating out the packages into a separate repo, let's say we need to make a major change that affects many packages
> e.g. we've spoken about requiring virtuals to have an explicit package, what if we didn't think of that until after separating out the package repo?

This is why we should probably have a versioned interface to the build_systems going forward, so that one recipe can
inherit from Makefile_v2 and another from Makefile_v3.   When we make a breaking change in code the recipes use, we 
then need to provide a backwards_compatible interface for recipes using the old interface, with suitable defaults (which 
will also discourage such changes, but *I* think that's a Good Thing).

**Comment by haampie** [comment]
With regard to compat between spack core and repositories: it should not block fixing issues like https://github.com/spack/spack/issues/46323 which require significant code shuffling and are unlikely to be completed before the 1.0 release.

So, practically speaking, `package.py` files should do only one import from spack core, namely `spack.package`, which already is a layer of indirection, allowing us to move functions around without causing breaking changes. The only thing that matters then is stable function signatures. Also we should explicitly re-export what we want to expose, instead re-exporting `*`.

**Comment by tgamblin** [comment]
> This is why we should probably have a versioned interface to the build_systems going forward, so that one recipe can
inherit from `Makefile_v2` and another from `Makefile_v3`.

For the first cut of this I think we are *just* trying to:
* Get the build systems into the `builtin` (I guess we should call it something else once it's separate) repo
* Allow people to depend on revisions/commits of that repo, e.g.:
  ```yaml
  repos:
      - url: https://github.com/spack/spack-pkgs
        ref: v1.0
  ```

Which gives you some amount of versioning already.  Whether we version the build systems inside that repo (or factor them out into their own repo and add inter-repo deps, which seems possibly more reasonable) is honestly more than I think we could take on for 1.0.  I *think* this will at least get us to a better place than we are now in 1.0.

Note also that with separate repos and with the build systems existing as something *in* a repo, I think you could add/override anything in the builtin repo with a local copy of a build system if needed -- so that avenue would be available as well, even if we don't formally version build systems.

**Comment by kwryankrattiger** [comment]
---
For the Github Actions CI:

There should be some kind of contract testing in both directions for the Spack Repo and Spack Core projects. I don't think those are mutually exclusive tests as the two repositories will certainly move at different rates and potentially have largely decoupled versioning schemes (TBD). I am not sure there is sufficient coverage if the compatibility checks are only done in one direction.

Minimally the contract should be running a package audit and concretization against some set of engineered environments that give reasonable results. Those environments could feasibly exist in an external repo in Spack used for defining CI targets. See below for other reasons to do that.

---
For Tracking the default repo:

The proposals to use a git urls/refs in the `repos.yaml` is by far my favorite. I would propose going a step further with what is currently proposed for tracking the `ref` and suggest that remote repos could follow the same type of model as the `version()` directive used in `package.py`. I think this would work well with the idea of having a `spack update <repo> <version>`, which could just change the one field in the repos.yaml config and override the local cached repo.

```yaml
repos:
  - git_url: https://github.com/spack/spack-pkgs
    url: https://github.com/spack/spack-pkgs/releases/download/v1.0.0/spack-1.0.0.tar.gz
    version: 1.0.0 # or git:develop
```

Tracking and updating submodules is a nightmare that I would very much prefer to avoid at all reasonable costs. Having worked on a number of different teams with varying experience working with Git repos, adding submodules is a sure fire way to frustrate some developers away from working on a project at all.

---
For the build caches:

We are now changing the definition of what matters for a release/develop build cache. Do we have a different pipeline that runs for each combination of Spack Core version and Spack Repo version? Or do we only track caches for one of those dimensions? As we have discussed previously, core changes in `asp.py` and the solver lp files can change package hashes as much as changes to `package.py` files, so how do we reconcile that in an meaningful way?

For the PR CI that generates build caches there is going to need to be some discussion on what should trigger CI. Right now we have a stateless query that tracks branches and pushes merge commits. Are now going to have to track multiple repos changing and create merge commits for each that run some CI? Do we need to test the Spack Repo changes against a sweep of "supported" Spack Core versions and visa-versa? The fact that these two repos will now float more or less independently and we want to provide some guarantees about the stability we should probably be testing that as much as is reasonable.

I don't think this is "solved" by using submodules since an update to the Spack Repo is still going to be delayed on testing until it is merged into the Spack Core. Using the URL approach, it is already possible to effectively do the same thing but without submodules.

The absolute simplest solution would be to completely ignore changes to hashes created by Spack Core and run the CI to create the build cache based entirely on the Spack Repo project and some Spack Core commit. And to keep some level of coverage in Spack Core CI it may be enough to just have the contract tests and leave populating the develop cache to the Spack Repo project. This opens up a set of issues we are partially covering with the current CI infrastructure where core and package changes interact unexpectedly, but maybe that is something to address later.

Otherwise we can look at setting up some triggers to funnel PR CI from the two repos into a common CI repo pipeline. I have some ideas on how to do that but it is quite a bit more complicated.

In either case, I think we would have to be careful on how this is done. It is probable based on how long CI takes to run that the `develop` branch of the other repo will move as CI is running, so we will minimally have to find a way to pin versions in CI. I think designing how the build farm CI will work should be considered closely with how we break things up.

---
Migration Procedure

I may have missed it somewhere, but we should be sure to maintain the git history from the current project for builtin repo packages to the new project. It would be a shame to lose that. I know of [BFG](https://github.com/rtyley/bfg-repo-cleaner) which can be used to quickly filter out files from git history. Do we trim down the git history in `spack/spack` while we are at it to remove every reference to the buildin repo? And/or just strip away the core and push that as the new Spack Repo project? If we filter the history from the `spack/spack` repo I imagine that will be a stark transition for many, the plus side of that though is the Spack Core repo gets a lot smaller, otherwise it will carry the legacy of host the builtin package repo forever.

The other thing to think about is what to do with all of the open issues/PRs into the main `spack/spack` repo and how to migrate those over to the new repo. I highly doubt we will close every single open package PR in the next 4 months. Is this going to be a case of figure it out your self, or will we do something with spackbot?

---
Where to put `build_systems`

I have mixed feelings about pulling `build_systems` into the Spack Repo. Glancing through the thread here I think most people want to put them into the Spack Repo project. My main concern is this will _require_ people to install the default repo to get access to these meta packages or, more likely, people will just start copy-pasting them around wholesale which I don't even want to start thinking about debugging.

**Comment by psakievich** [comment]
-----

One of the deployment mechansims that has been on my mind for this would be @haampie 's prototype of using zip files for the packages. I searched for the issue or PR but couldn't find so I'll like later if I find it.  If the production mainline method is to pull in a zip file per repo then that could ease some of these deployment and linking repos. We would still need to support having them in unzipped/git form, but I think if we had a default that used a zip file then we could avoid projecting git submodules or other more complex vendoring methods on to users as the default. 

Getting the repo would essentially fall into the same mechanism we develop for bootstrapping. Perhaps there is a meta package for package repo that mainly handles the version checking and fetching protocols. This would recycle a lot of our current methodology.

----
Another potential break here that hasn't been noted yet is all of the people using package inheritance.  This has become a defacto standard for how to stay synced with mainline spack rather than forking, but still have custom behavior or adaptions.  During this refactor I would like to see this become a first class citizen of Spack. We should add testing and documentation on this.

I think this could double as the method for dealing with @kwryankrattiger's concerns on the `build_systems` copy and deploy etc.  If we made `build_systems` basically a smaller repo that is required, people could extend/add new ones as needed.

----

**Comment by kwryankrattiger** [comment]
> Getting the repo would essentially fall into the same mechanism we develop for bootstrapping.

I started thinking this way, but it felt a little strange to me to when I started going through the workflow.

To start it is fine, makes perfect sense,
```console
spack bootstrap
spack spec zlib
```

But then upgrading and managing the `spack-pkgs` version, how do you do that with bootstrap? I don't see an `update` function there, just `bootstrap` and then you are done, never think about what was bootstrapped again!

**Comment by psakievich** [comment]
> But then upgrading and managing the `spack-pkgs` version, how do you do that with bootstrap? I don't see an `update` function there, just `bootstrap` and then you are done, never think about what was bootstrapped again!

I think we'd need to add that.  Probably the along the lines of @scheibelp 's suggestion for a `spack update` or a `spack package update`.  If we consider the package repo will have multiple releases then we need a semantic for what is installed, what is compatible, how to install further.  I don't think it should be considered an addition to the bootstrapping command. Just first go we have bootstrapping pull what is the default for the cloned instance of spack, but then going forward you'd need a way to update and grab things regardless. No matter what we do I don't see how we avoid this need for update and compatibility checking.

 I like the idea of the zip deployment because

1. Performance implications for filesystem (1 file vs 8k and growing)
2. Mirrors our other fetch mechanisms (pull a tar ball)

2) is the key here for me in that having a single zipped, checksummable file makes things easier to manage and update quickly.  If we do git shenanigans then we have a lot more issues with potential for `builtin.pkg.foo` to do weird things. Which we can allow but say is not the production path kind of how we deal with git versions over sha256 for package versions.

**Comment by haampie** [comment]
I've done some basic testing regarding the change of directory structure of the builtin repo in `spack/spack-packages`, and looks like transitioning is not a major issue. Over there:

* `var/spack/repos/builtin/` is renamed to `spack_repo/builtin/` where `spack_repo` serves as a namespace module;
* git history is preserved;
* it's possible to cherry-pick commits from PRs against `spack/spack` without any problem.

```
$ git clone https://github.com/spack/spack-packages

$ cd spack-packages

$ tree -L 3
.
└── spack_repo
    └── builtin
        ├── packages
        └── repo.yaml

$ git fetch https://github.com/spack/spack.git pull/49019/head

$ git cherry-pick FETCH_HEAD
[develop 44647e7455] magma: remove cuda_arch constraint on 2.9.0+
...
 1 file changed, 5 insertions(+), 4 deletions(-)
```

**Comment by haampie** [comment]
This list of imports remains to be fixed in builtin:

https://gist.github.com/haampie/297680f55b8f216e1fd3de466d19346d

**Comment by dylex** [comment]
Would it be possible to share the exact commands (BFG or git-filter-branch or otherwise) that were used to create the spack-packages repo from spack? We are finally trying to merge changes into a spack fork, and it would be nice to be able to recreate this in order to preserve history of our changes in the merge. (Apologies if I missed this somewhere.)

**Comment by kwryankrattiger** [comment]
To migrate PRs to spack-packages you can use the tool https://github.com/spack/migrate-package-prs

If you run into issues you can manually cherry-pick the commits from the PR ref and resolve the conflicts. There should be no need to do filter branch.

**Comment by tgamblin** [comment]
@dylex: If you really want to see how it was done, look at:
* https://github.com/spack/spack/pull/50322
* https://github.com/spack/spack/pull/50791
* https://github.com/spack/spack/pull/50804

The migration wasn't immediate -- we started by constantly syncing github.com/spack/spack to github.com/spack/spack-packages, then got `spack` working with `spack-packages` as an external repo, then finally cut over completely once the internal repo was no longer used.  The sync used [`git-filter-repo`](https://github.com/newren/git-filter-repo/), which is actually recommended amid a slew of warnings in the [`git filter-branch` docs](https://github.com/newren/git-filter-repo/).  You can see the actual commands in the GitHub action in the first PR above.

If you are already working on a clone of `spack` and have changes there, I think you can keep them, but cherry-pick the changes onto `spack-packages` like @kwryankrattiger suggests above. I think that will be simpler than trying to filter.  Also, I do not think you need to filter since you can add `spack/spack` (or a clone of it) as a remote in a `spack/spack-packages` repository and cherry pick -- `spack` commits should apply cleanly to `spack-packages` unless you've introduced your own conflicts.
